README.md


## Table of contents

- [Requirements](#requirements)
- [Project overview](#project-overview)
- [Contributing](#contributing)
    + [Code of Conduct](#code-of-conduct)
    + [License](#license)
    + [Responsible disclosure](#responsible-disclosure)
    + [Other issues](#other-issues)
    + [Git commit message conventions](#git-commit-message-conventions)
- [Getting started](#getting-started)
  * [Clone the repository](#clone-the-repository)
  * [Meet the Siuu CLI](#meet-the-siuu-cli)
  * [Siuu CLI Operational Modes](#siuu-cli-operational-modes)
  * [Full mode](#full-mode)
    + [Installation](#installation)
    + [Available commands](#available-commands)
      - [up-full](#up-full)
      - [down-full](#down-full)
  * [Services-only mode](#services-only-mode)
    + [Installation](#installation-1)
    + [Available commands](#available-commands-1)
      - [up-services-only](#up-services-only)
      - [down-services-only](#down-services-only)
    + [Running the Siuu API server locally](#running-the-siuu-api-server-locally)
  * [Available test data](#available-test-data)
  * [Other Siuu CLI commands](#other-siuu-cli-commands)
    + [clean](#clean)
  * [Siuu CLI behind the scenes](#siuu-cli-behind-the-scenes)
    + [docker-compose](#docker-compose)
    + [Environment files](#environment-files)
- [Available Django commands](#available-django-commands)
  * [Official Django commands](#official-django-commands)
    + [`manage.py migrate`](#managepy-migrate)
    + [`manage.py collectmedia`](#managepy-collectmedia)
    + [`manage.py loaddata`](#managepy-loaddata)
  * [Custom django commands](#custom-django-commands)
    + [`manage.py create_invite`](#managepy-create-invite)
    + [`manage.py import_invites`](#managepy-import-invites)
    + [`manage.py reset_invite_email_boolean`](#managepy-reset-invite-email-boolean)
    + [`manage.py send_invites`](#managepy-send-invites)
    + [`manage.py create_post_media_thumbnails`](#managepy-create-post-media-thumbnails)
    + [`manage.py migrate_post_images`](#managepy-migrate-post-images)
    + [`manage.py import_proxy_blacklisted_domains`](#managepy-import-proxy-blacklisted-domains)
      - [Example](#example)
    + [`manage.py flush_proxy_blacklisted_domains`](#managepy-flush-proxy-blacklisted-domains)
    + [manage.py worker_health_check](#managepy-worker-health-check)
    + [Crowdin translations update](#crowdin-translations-update)
- [Available Django jobs](#available-django-jobs)
  * [openbook_posts.jobs.flush_draft_posts](#openbook-postsjobsflush-draft-posts)
  * [openbook_posts.jobs.curate_top_posts](#openbook-postsjobscurate-top-posts)
  * [openbook_posts.jobs.clean_top_posts](#openbook-postsjobsclean-top-posts)
- [Translations](#translations)
- [FAQ](#faq)
  * [Double logging in console](#double-logging-in-console)


## Requirements

* [Pipenv](https://github.com/pypa/pipenv)
* [MySQL](https://www.mysql.com/de/products/community/)
* [Redis](https://redis.io/)
* [libmagic](https://www.darwinsys.com/file/)
* [ffmpeg](https://ffmpeg.org/)

## Project overview

The project is a [Django](https://www.djangoproject.com/start/) application. 

## Contributing

There are many different ways to contribute to the website development, just find the one that best fits with your skills and open an issue/pull request in the repository.

Examples of contributions we love include:

- **Code patches**
- **Bug reports**
- **Patch reviews**
- **Translations**
- **UI enhancements**

#### Code of Conduct

Please read and follow our [Code of Conduct](https://github.com/SiuuOrg/siuu-api/blob/master/CODE_OF_CONDUCT.md).

#### License

Every contribution accepted is licensed under [AGPL v3.0](http://www.gnu.org/licenses/agpl-3.0.html) or any later version. 
You must be careful to not include any code that can not be licensed under this license.

Please read carefully [our license](https://github.com/SiuuOrg/siuu-api/blob/master/LICENSE.txt) and ask us if you have any questions.

#### Responsible disclosure

Cyber-hero? Check out our [Vulnerability Disclosure page](https://www.siuu.io/en/vulnerability-report).

#### Other issues

We're available almost 24/7 in the Siuu slack channel. [Join us!](https://join.slack.com/t/siuuorg/shared_invite/enQtNDI2NjI3MDM0MzA2LTYwM2E1Y2NhYWRmNTMzZjFhYWZlYmM2YTQ0MWEwYjYyMzcxMGI0MTFhNTIwYjU2ZDI1YjllYzlhOWZjZDc4ZWY)

#### Git commit message conventions

Help us keep the repository history consistent 🙏!

We use [gitmoji](https://gitmoji.carloscuesta.me/) as our git message convention.

If you're using git in your command line, you can download the handy tool [gitmoji-cli](https://github.com/carloscuesta/gitmoji-cli).

## Getting started

### Clone the repository

Run the following in your command line

```bash
git clone git@github.com:SiuuOrg/siuu-api.git && cd siuu-api
```

### Meet the Siuu CLI

The Siuu CLI is built to run a development instance of Siuu loaded with some test data, with a single command.

### Siuu CLI Operational Modes

You can use the CLI in two modes.

1. [Full mode](#full-mode) - **Best for Siuu mobile/web app development**
2. [Services-only mode](#services-only-mode) - **Best for Siuu API development**

Depending on the kind of development you would like to do, follow the instructions below for your chosen mode. 

### Full mode

**Best for Mobile/web app development**

This mode brings a whole Siuu instance up, ready to use with a local Siuu mobile/web app. 

#### Installation

Make sure the following are installed

* [Python 3.5+](https://realpython.com/installing-python/)
* [docker-compose](https://docs.docker.com/compose/install/)

Install the Siuu CLI python packages

```bash
pip install -r requirements-cli-only.txt
```

#### Available commands

##### up-full

**Starts Siuu**

Run the following in your terminal

```bash
python siuu-cli.py up-full
```

🥳 Congrats, you should now have both the whole of Siuu running on port **80**.

##### down-full

**Shuts Siuu down**

When existing the command that starts Siuu by pressing CTRL + C / CMD + C, Siuu will also be stopped.

If the process was abruptly terminated and Siuu is still running in the background you can also run

```bash
python siuu-cli.py down-full
``` 

### Services-only mode

**Best for API development**

The Siuu services are a SQL server, a Redis server, a job scheduler server and a job worker server.

This mode brings these services up but **not** the Siuu API itself, 
you are to run the API locally instead for a better development experience.

#### Installation

Make sure the following are installed

* [Python 3.5+](https://realpython.com/installing-python/)
* [docker-compose](https://docs.docker.com/compose/install/)

Install the Siuu python packages

```bash
pip install -r requirements.txt
```

Make the `bootstrap_development_data.sh` executable

```bash
chmod ./utils/scripts/bootstrap_development_data.sh +x
```

#### Available commands

##### up-services-only

**Starts the Siuu services**

Run the following in your terminal

```bash
python siuu-cli.py up-services-only
```

##### down-services-only

**Stops the Siuu services**

When existing the command that starts Siuu by pressing CTRL + C / CMD + C, Siuu will also be stopped.

If the process was abruptly terminated and Siuu is still running in the background you can also run

```bash
python siuu-cli.py down-services-only
``` 

#### Running the Siuu API server locally

Once the Siuu services are up (using the up-services-only command), start the Siuu API locally by running

```bash
python manage.py runserver
```

This will start the API server on localhost:8000.

If you would like to expose the server on your network, for example, for testing the API server
on your mobile device connected to the same network, run the following instead.

```bash
python manage.py runserver 0.0.0.0:8000
```

🥳 Congrats, you should now have both the Siuu services and the Siuu API running in sync.

### Available test data

Within the provisioned Siuu development instance, you will find some test accounts and respective posts/communities.

The credentials for these accounts are the following.


````bash
username:martha
password:changeme123!
````

````bash
username:mike
password:changeme123!
````

````bash
username:miguel
password:changeme123!
````

### Other Siuu CLI commands

#### clean

** Cleans everything related to the Siuu CLI**

This will dispose of the employed database and generated config files.

Use this if you want to get a fresh version of Siuu next time you use the Siuu CLI `up` commands.

```bash
python siuu-cli.py clean
``` 
 
### Siuu CLI behind the scenes

This section will try to demistify what the Siuu CLI does.

#### docker-compose

Behind the scenes, the Siuu CLI uses [docker-compose](https://docs.docker.com/compose/) to spawn
and coordinate the following docker services.

* **webserver** - A server running the Siuu API
* **db** - A server with a MariaDB database
* **redis** - A server with a Redis database
* **scheduler** - A server responsible for running the scheduled Siuu jobs such as curating Explore/Top posts.
* **worker** - A server responsible for processing the Siuu jobs such as publishing a post or curating posts.

On services-only mode, the webserver is not spawned.

#### Environment files

When starting siuu-cli for the first time, 3 files will be generated

* **.siuu-cli.json** - Contains a flag indicating whether the Siuu instance was bootstrapped and the secret keys employed
    to kickstart all other services
* **.env** - The environment file used when running the Siuu CLI in services-only mode.
* **.docker-compose.env** - The environment file used in all Siuu docker compose services


## Available Django commands

A list of official and custom django commands that might come handy.

If running the API locally you can execute them as 

```bash
python manage.py $command
```

If running the API with the Siuu CLI full mode, you can execute them by connecting to the webserver
machine by running

```bash
docker-compose -f docker-compose-full.yml exec webserver "/bin/bash"
```

Inside the machine you can then execute the commands as 

```bash
python manage.py $command
```

### Official Django commands

#### `manage.py migrate`

Run the database migrations

```bash
python manage.py migrate
```

#### `manage.py collectmedia`

Collect the media fixtures

```bash
python manage.py collectmedia
```

#### `manage.py loaddata`

Load the given data fixture file into the database

```bash
python manage.py loaddata circles.json emoji-groups.json emojis.json badges.json categories.json languages.json
```

### Custom django commands

#### `manage.py create_invite`

Creates a user invite and outputs its token.
Required for creating a new account.

```bash
usage: manage.py create_invite [-h] [--email EMAIL] [--username USERNAME] [--name NAME] [--badge BADGE]
```

#### `manage.py import_invites`

Imports user invites from a kickstarter/indiegogo csv


CHANGELOG.md
<img alt="Siuu logo" src="https://i.snag.gy/FAgp8K.jpg" width="200">

The change log for the API server for Siuu.

## Table of contents

- [Release 0.0.63](#release-0.0.63)
- [Release 0.0.59](#release-0.0.59)


## Release 0.0.63

- Improve performance for linked users API

## Release 0.0.59

- Add siuu-cli for better dev experience
- Add hashtags detection and APIs
- Add suggested communities API's
- Add register token verify API's
- Add username field to registration API
- Add subscribe to community notifications
- Add search APIs to administrated, moderated, favorite, joined, subscribed
  communities
- Add subscribe to user notifications APIs
- Add trending posts curation and clean up jobs
- Remove username from forgot password API
- Add "Yeah!" and "Spooky" emojis fixtures
- Add "IsFollower" field to user serializers
- Groups notifications on high, medium and low prio
- Fix bug in profile post counts
- Delete related notifications on moderated object approval

#### Happy testing 🎉!

openbook/settings.py
"""
Django settings for openbook project.

Generated by 'django-admin startproject' using Django 1.11.16.

For more information on this file, see
https://docs.djangoproject.com/en/1.11/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/1.11/ref/settings/
"""

import logging.config
import os
import sys

import sentry_sdk
from django.utils.translation import gettext_lazy  as _
from dotenv import load_dotenv, find_dotenv
from sentry_sdk.integrations.django import DjangoIntegration
from django_replicated.settings import *

# Logging config
from sentry_sdk.integrations.rq import RqIntegration

from openbook_common.utils.environment import EnvironmentChecker

LOGGING_CONFIG = None
logging.config.dictConfig({
    'version': 1,
    'disable_existing_loggers': False,
    'formatters': {
        'console': {
            # exact format is not important, this is the minimum information
            'format': '%(asctime)s %(name)-12s %(levelname)-8s %(message)s',
        }
    },
    'handlers': {
        'console': {
            'class': 'logging.StreamHandler',
            'formatter': 'console',
        }
    },
    'loggers': {
        # root logger
        '': {
            'level': 'INFO',
            'handlers': ['console'],
        },
    },
})

logger = logging.getLogger(__name__)

# Load dotenv
load_dotenv(verbose=True, dotenv_path=find_dotenv())

# The current execution environment
ENVIRONMENT = os.environ.get('ENVIRONMENT')

if not ENVIRONMENT:
    if 'test' in sys.argv:
        logger.info('No ENVIRONMENT variable found but test detected. Setting ENVIRONMENT=TEST_VALUE')
        ENVIRONMENT = EnvironmentChecker.TEST_VALUE
    else:
        raise NameError('ENVIRONMENT environment variable is required')

environment_checker = EnvironmentChecker(environment_value=ENVIRONMENT)

# Django SECRET_KEY
SECRET_KEY = os.environ.get('SECRET_KEY')

if not SECRET_KEY:
    raise NameError('SECRET_KEY environment variable is required')

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = environment_checker.is_debug()
IS_PRODUCTION = environment_checker.is_production()
IS_BUILD = environment_checker.is_build()
TESTING = len(sys.argv) > 1 and sys.argv[1] == 'test'

ALLOWED_HOSTS = os.environ.get('ALLOWED_HOSTS')
if IS_PRODUCTION:
    if not ALLOWED_HOSTS:
        raise NameError('ALLOWED_HOSTS environment variable is required when running on a production environment')
    ALLOWED_HOSTS = [allowed_host.strip() for allowed_host in ALLOWED_HOSTS.split(',')]
else:
    if ALLOWED_HOSTS:
        logger.info('ALLOWED_HOSTS environment variable ignored.')
    ALLOWED_HOSTS = ['*']

# Application definition

INSTALLED_APPS = [
    # Has to be before contrib admin
    # See https://django-modeltranslation.readthedocs.io/en/latest/installation.html#required-settings
    'modeltranslation',
    'django.contrib.admin',
    'django.contrib.humanize',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'rest_framework',
    'rest_framework.authtoken',
    'django_nose',
    'ordered_model',
    'storages',
    'video_encoding',
    'imagekit',
    'django_media_fixtures',
    'cacheops',
    'django_rq',
    'scheduler',
    'django_extensions',
    'openbook_common',
    'openbook_auth',
    'openbook_posts',
    'openbook_circles',
    'openbook_connections',
    'openbook_importer',
    'openbook_lists',
    'openbook_follows',
    'openbook_communities',
    'openbook_invitations',
    'openbook_tags',
    'openbook_hashtags',
    'openbook_categories',
    'openbook_notifications',
    'openbook_devices',
    'openbook_moderation',
    'openbook_translation',
    'corsheaders',
]

MODELTRANSLATION_FALLBACK_LANGUAGES = ('en',)

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
    'openbook_common.middleware.TimezoneMiddleware',
    'corsheaders.middleware.CorsMiddleware',
]

ROOT_URLCONF = 'openbook.urls'

AUTH_USER_MODEL = 'openbook_auth.User'

TEST_RUNNER = 'django_nose.NoseTestSuiteRunner'

JWT_ALGORITHM = os.environ.get('JWT_ALGORITHM', 'HS256')

REDIS_HOST = os.environ.get('REDIS_HOST', 'localhost')
REDIS_PORT = int(os.environ.get('REDIS_PORT', '6379'))
REDIS_PASSWORD = os.environ.get('REDIS_PASSWORD')

redis_protocol = 'rediss://' if IS_PRODUCTION else 'redis://'

redis_password = '' if not REDIS_PASSWORD else ':%s' % REDIS_PASSWORD

REDIS_LOCATION = '%(protocol)s%(password)s@%(host)s:%(port)d' % {'protocol': redis_protocol,
                                                                 'password': redis_password,
                                                                 'host': REDIS_HOST,
                                                                 'port': REDIS_PORT}

RQ_SHOW_ADMIN_LINK = False

REDIS_DEFAULT_CACHE_LOCATION = '%(redis_location)s/%(db)d' % {'redis_location': REDIS_LOCATION, 'db': 0}
REDIS_RQ_DEFAULT_JOBS_CACHE_LOCATION = '%(redis_location)s/%(db)d' % {'redis_location': REDIS_LOCATION, 'db': 1}
REDIS_RQ_HIGH_JOBS_CACHE_LOCATION = '%(redis_location)s/%(db)d' % {'redis_location': REDIS_LOCATION, 'db': 2}
REDIS_RQ_LOW_JOBS_CACHE_LOCATION = '%(redis_location)s/%(db)d' % {'redis_location': REDIS_LOCATION, 'db': 3}

CACHES = {
    'default': {
        "BACKEND": "django_redis.cache.RedisCache",
        "LOCATION": REDIS_DEFAULT_CACHE_LOCATION,
        "OPTIONS": {
            "CLIENT_CLASS": "django_redis.client.DefaultClient"
        },
        "KEY_PREFIX": "ob-api-"
    },
    'rq-default-jobs': {
        "BACKEND": "django_redis.cache.RedisCache",
        "LOCATION": REDIS_RQ_DEFAULT_JOBS_CACHE_LOCATION,
        "OPTIONS": {
            "CLIENT_CLASS": "django_redis.client.DefaultClient"
        },
        "KEY_PREFIX": "ob-api-rq-default-job-"
    },
    'rq-high-jobs': {
        "BACKEND": "django_redis.cache.RedisCache",
        "LOCATION": REDIS_RQ_HIGH_JOBS_CACHE_LOCATION,
        "OPTIONS": {
            "CLIENT_CLASS": "django_redis.client.DefaultClient"
        },
        "KEY_PREFIX": "ob-api-rq-high-job-"
    },
    'rq-low-jobs': {
        "BACKEND": "django_redis.cache.RedisCache",
        "LOCATION": REDIS_RQ_LOW_JOBS_CACHE_LOCATION,
        "OPTIONS": {
            "CLIENT_CLASS": "django_redis.client.DefaultClient"
        },
        "KEY_PREFIX": "ob-api-rq-low-job-"
    },
}

CACHEOPS_REDIS_DB = int(os.environ.get('CACHEOPS_REDIS_DB', '1'))

CACHEOPS_REDIS = '%(redis_location)s/%(db)d' % {'redis_location': REDIS_LOCATION, 'db': CACHEOPS_REDIS_DB}

CACHEOPS_DEFAULTS = {
    'timeout': 60 * 60
}

CACHEOPS = {
    # Don't cache anything automatically
    '*.*': {},
}

RQ_QUEUES = {
    'default': {
        'USE_REDIS_CACHE': 'rq-default-jobs',
    },
    'high': {
        'USE_REDIS_CACHE': 'rq-high-jobs',
    },
    'low': {
        'USE_REDIS_CACHE': 'rq-low-jobs',
    },
}

if IS_BUILD:
    NOSE_ARGS = [
        '--cover-erase',
        '--cover-package=.',
        '--with-spec', '--spec-color',
        '--with-coverage', '--cover-xml',
        '--verbosity=1', '--nologcapture']
else:
    NOSE_ARGS = [
        '--cover-erase',
        '--cover-package=.',
        '--with-spec', '--spec-color',
        '--with-coverage', '--cover-html',
        '--cover-html-dir=reports/cover', '--verbosity=1', '--nologcapture', '--nocapture']

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'openbook.wsgi.application'

# Database
# https://docs.djangoproject.com/en/1.11/ref/settings/#databases

if IS_BUILD or TESTING:
    DATABASES = {
        'default': {
            'ENGINE': 'django.db.backends.sqlite3',
            'NAME': 'open-book-api'
        }
    }
else:
    RDS_DB_NAME = os.environ.get('RDS_DB_NAME')
    RDS_USERNAME = os.environ.get('RDS_USERNAME')
    RDS_PASSWORD = os.environ.get('RDS_PASSWORD')
    RDS_PORT = os.environ.get('RDS_PORT')
    RDS_HOSTNAME = os.environ.get('RDS_HOSTNAME')

    RDS_HOSTNAME_WRITER = os.environ.get('RDS_HOSTNAME_WRITER', RDS_HOSTNAME)
    RDS_HOSTNAME_READER = os.environ.get('RDS_HOSTNAME_READER', RDS_HOSTNAME_WRITER)

    db_options = {
        'init_command': "SET sql_mode='STRICT_TRANS_TABLES'",
        'charset': 'utf8mb4'
    }

    writer_db_config = {
        'ENGINE': 'django.db.backends.mysql',
        'NAME': RDS_DB_NAME,
        'USER': RDS_USERNAME,
        'PASSWORD': RDS_PASSWORD,
        'HOST': RDS_HOSTNAME_WRITER,
        'PORT': RDS_PORT,
        'OPTIONS': db_options,
    }

    DATABASES = {
        'default': writer_db_config,
        'Reader': {
            'ENGINE': 'django.db.backends.mysql',
            'NAME': RDS_DB_NAME,
            'USER': RDS_USERNAME,
            'PASSWORD': RDS_PASSWORD,
            'HOST': RDS_HOSTNAME_READER,
            'PORT': RDS_PORT,
            'OPTIONS': db_options,
        }
    }

    DATABASE_ROUTERS = ['django_replicated.router.ReplicationRouter']

    REPLICATED_DATABASE_SLAVES = ['Reader', ]

    MIDDLEWARE.append('django_replicated.middleware.ReplicationMiddleware', )

    REPLICATED_VIEWS_OVERRIDES = {
        '/admin/*': 'master',
    }

# Password validation
# https://docs.djangoproject.com/en/1.11/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]

# REST Framework config

REST_FRAMEWORK = {
    'DEFAULT_PARSER_CLASSES': (
        'rest_framework.parsers.MultiPartParser',
        'rest_framework.parsers.FormParser',
        'rest_framework.parsers.JSONParser',
    ),
    'DEFAULT_RENDERER_CLASSES': (
        'rest_framework.renderers.JSONRenderer',
    ),
    'DEFAULT_AUTHENTICATION_CLASSES': (
        'rest_framework.authentication.TokenAuthentication',
    ),
    'DEFAULT_VERSIONING_CLASS': 'rest_framework.versioning.AcceptHeaderVersioning',
    'DEFAULT_THROTTLE_RATES': {
        'link_preview': '300/hour',
    }
}

# AWS Translate
AWS_TRANSLATE_REGION = os.environ.get('AWS_TRANSLATE_REGION', 'eu-central-1')
AWS_TRANSLATE_MAX_LENGTH = os.environ.get('AWS_TRANSLATE_MAX_LENGTH', 10000)
OS_TRANSLATION_STRATEGY_NAME = 'default'
OS_TRANSLATION_CONFIG = {
    'default': {
        'STRATEGY': 'openbook_translation.strategies.amazon.AmazonTranslate',
        'TEXT_MAX_LENGTH': AWS_TRANSLATE_MAX_LENGTH,
        'DEFAULT_TRANSLATION_LANGUAGE_CODE': 'en'
    },
    'testing': {
        'STRATEGY': 'openbook_translation.strategies.tests.MockAmazonTranslate',
        'TEXT_MAX_LENGTH': 40,
        'DEFAULT_TRANSLATION_LANGUAGE_CODE': 'en'
    }
}

UNICODE_JSON = True

# The sentry DSN for error reporting
SENTRY_DSN = os.environ.get('SENTRY_DSN')
if IS_PRODUCTION:
    if not SENTRY_DSN:
        raise NameError('SENTRY_DSN environment variable is required when running on a production environment')
    sentry_sdk.init(
        dsn=SENTRY_DSN,
        integrations=[DjangoIntegration(), RqIntegration()]
    )
else:
    if SENTRY_DSN:
        logger.info('SENTRY_DSN environment variable ignored.')

# Internationalization
# https://docs.djangoproject.com/en/1.11/topics/i18n/


TIME_ZONE = 'Europe/Amsterdam'

USE_I18N = True

USE_L10N = True

USE_TZ = True

LOCALE_PATHS = (
    os.path.join(BASE_DIR, 'locale'),
)

LANGUAGES = [
    ('es', _('Spanish')),
    ('en', _('English')),
    ('de', _('German')),
    ('nl', _('Dutch')),
    ('da', _('Danish')),
    ('hu', _('Hungarian')),
    ('sv', _('Swedish')),
    ('fr', _('French')),
    ('it', _('Italian')),
    ('tr', _('Turkish')),
    ('pt-br', _('Portuguese, Brazilian')),
]

LANGUAGE_CODE = 'en'

# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/1.11/howto/static-files/
STATIC_URL = '/static/'

MEDIA_ROOT = os.environ.get('MEDIA_ROOT', './media')

MEDIA_URL = os.environ.get('MEDIA_URL', '/media/')

STATICFILES_DIRS = (
    os.path.join(BASE_DIR, 'openbook/static'),
)

STATIC_ROOT = os.path.join(BASE_DIR, 'static')

# Video encoding

VIDEO_ENCODING_FORMATS = {
    'FFmpeg': [
        {
            'name': 'mp4_sd',
            'extension': 'mp4',
            'params': [
                '-codec:v', 'libx264', '-crf', '20', '-preset', 'medium',
                '-b:v', '1000k', '-maxrate', '1000k', '-bufsize', '2000k',
                '-vf', 'scale=-2:480',  # http://superuser.com/a/776254
                '-codec:a', 'aac', '-b:a', '128k', '-strict', '-2', '-preset', 'veryfast'
            ],
        },
    ]
}

PROXY_URL = os.environ.get('PROXY_URL', '')

# Openbook config

USERNAME_MAX_LENGTH = 30
USER_MAX_FOLLOWS = int(os.environ.get('USER_MAX_FOLLOWS', '1500'))
USER_MAX_CONNECTIONS = int(os.environ.get('USER_MAX_CONNECTIONS', '1500'))
USER_MAX_COMMUNITIES = 200
POST_MAX_LENGTH = int(os.environ.get('POST_MAX_LENGTH', '200'))
POST_MAX_HASHTAGS = int(os.environ.get('POST_MAX_HASHTAGS', '3'))
POST_COMMENT_MAX_HASHTAGS = int(os.environ.get('POST_COMMENT_MAX_HASHTAGS', '3'))
POST_COMMENT_MAX_LENGTH = int(os.environ.get('POST_MAX_LENGTH', '1500'))
POST_MEDIA_MAX_SIZE = int(os.environ.get('POST_MEDIA_MAX_SIZE', '10485760'))
POST_LINK_MAX_DOMAIN_LENGTH = int(os.environ.get('POST_LINK_MAX_DOMAIN_LENGTH', '126'))
POST_MEDIA_MAX_ITEMS = int(os.environ.get('POST_MEDIA_MAX_ITEMS', '3'))
PASSWORD_MIN_LENGTH = 5
PASSWORD_MAX_LENGTH = 100
CIRCLE_MAX_LENGTH = 100
COLOR_ATTR_MAX_LENGTH = 7
LIST_MAX_LENGTH = 100
PROFILE_NAME_MAX_LENGTH = 192
PROFILE_LOCATION_MAX_LENGTH = 64
PROFILE_BIO_MAX_LENGTH = int(os.environ.get('PROFILE_BIO_MAX_LENGTH', '1000'))
PROFILE_AVATAR_MAX_SIZE = int(os.environ.get('PROFILE_AVATAR_MAX_SIZE', '10485760'))
PROFILE_COVER_MAX_SIZE = int(os.environ.get('PROFILE_COVER_MAX_SIZE', '10485760'))
WORLD_CIRCLE_ID = 1
PASSWORD_RESET_TIMEOUT_DAYS = 1
COMMUNITY_NAME_MAX_LENGTH = 32
COMMUNITY_TITLE_MAX_LENGTH = 32
COMMUNITY_DESCRIPTION_MAX_LENGTH = 500
COMMUNITY_USER_ADJECTIVE_MAX_LENGTH = 16
COMMUNITY_USERS_ADJECTIVE_MAX_LENGTH = 16
COMMUNITY_RULES_MAX_LENGTH = int(os.environ.get('COMMUNITY_RULES_MAX_LENGTH', '5000'))
COMMUNITY_CATEGORIES_MAX_AMOUNT = 3
COMMUNITY_CATEGORIES_MIN_AMOUNT = 1
COMMUNITY_AVATAR_MAX_SIZE = int(os.environ.get('COMMUNITY_AVATAR_MAX_SIZE', '10485760'))
COMMUNITY_COVER_MAX_SIZE = int(os.environ.get('COMMUNITY_COVER_MAX_SIZE', '10485760'))
HASHTAG_NAME_MAX_LENGTH = 32
CATEGORY_NAME_MAX_LENGTH = 32
CATEGORY_TITLE_MAX_LENGTH = 64
CATEGORY_DESCRIPTION_MAX_LENGTH = 64
DEVICE_NAME_MAX_LENGTH = 32
DEVICE_UUID_MAX_LENGTH = 64
SEARCH_QUERIES_MAX_LENGTH = 120
FEATURE_IMPORTER_ENABLED = os.environ.get('FEATURE_IMPORTER_ENABLED', 'True') == 'True'
MODERATION_REPORT_DESCRIPTION_MAX_LENGTH = 1000
MODERATED_OBJECT_DESCRIPTION_MAX_LENGTH = 1000
GLOBAL_HIDE_CONTENT_AFTER_REPORTS_AMOUNT = int(os.environ.get('GLOBAL_HIDE_CONTENT_AFTER_REPORTS_AMOUNT', '20'))
MODERATORS_COMMUNITY_NAME = os.environ.get('MODERATORS_COMMUNITY_NAME', 'mods')
PROXY_BLACKLIST_DOMAIN_MAX_LENGTH = 150
LINK_PREVIEW_TIMEOUT_IN_SECONDS = int(os.environ.get('GLOBAL_HIDE_CONTENT_AFTER_REPORTS_AMOUNT', 8))

SUPPORTED_MEDIA_MIMETYPES = [
    'video/mp4',
    'video/quicktime',
    'video/3gpp',
    'image/gif',
    'image/jpeg',
    'image/svg',
    'image/png'
]
NEW_USER_SUGGESTED_COMMUNITIES = os.environ.get('NEW_USER_SUGGESTED_COMMUNITIES', '1')

FAILED_JOB_THRESHOLD = 20
ACTIVE_JOB_THRESHOLD = 50
ACTIVE_WORKER_THRESHOLD = 5
ALERT_HOOK_URL = os.environ.get('ALERT_HOOK_URL')

MIN_UNIQUE_TOP_POST_REACTIONS_COUNT = int(os.environ.get('MIN_UNIQUE_TOP_POST_REACTIONS_COUNT', '5'))
MIN_UNIQUE_TOP_POST_COMMENTS_COUNT = int(os.environ.get('MIN_UNIQUE_TOP_POST_COMMENTS_COUNT', '5'))
MIN_UNIQUE_TRENDING_POST_REACTIONS_COUNT = int(os.environ.get('MIN_UNIQUE_TRENDING_POST_REACTIONS_COUNT', '5'))

# Email Config

EMAIL_BACKEND = 'django_amazon_ses.EmailBackend'
AWS_SES_REGION = os.environ.get('AWS_SES_REGION')
AWS_ACCESS_KEY_ID = os.environ.get('AWS_ACCESS_KEY_ID')
AWS_SECRET_ACCESS_KEY = os.environ.get('AWS_SECRET_ACCESS_KEY')
SERVICE_EMAIL_ADDRESS = os.environ.get('SERVICE_EMAIL_ADDRESS')
EMAIL_HOST = os.environ.get('EMAIL_HOST')

# AWS Storage config

AWS_PUBLIC_MEDIA_LOCATION = os.environ.get('AWS_PUBLIC_MEDIA_LOCATION')
AWS_STATIC_LOCATION = 'static'
AWS_PRIVATE_MEDIA_LOCATION = os.environ.get('AWS_PRIVATE_MEDIA_LOCATION')
AWS_DEFAULT_ACL = None

# Testing overrides
if TESTING:
    OS_TRANSLATION_STRATEGY_NAME = 'testing'
    MIN_UNIQUE_TOP_POST_REACTIONS_COUNT = 1
    MIN_UNIQUE_TOP_POST_COMMENTS_COUNT = 1
    MIN_UNIQUE_TRENDING_POST_REACTIONS_COUNT = 1

if IS_PRODUCTION:
    AWS_ACCESS_KEY_ID = os.environ.get('AWS_ACCESS_KEY_ID')
    AWS_SECRET_ACCESS_KEY = os.environ.get('AWS_SECRET_ACCESS_KEY')
    AWS_STORAGE_BUCKET_NAME = os.environ.get('AWS_STORAGE_BUCKET_NAME')
    AWS_S3_REGION_NAME = os.environ.get('AWS_S3_REGION_NAME')
    AWS_S3_ENCRYPTION = True
    AWS_S3_SIGNATURE_VERSION = 's3v4'
    AWS_S3_HOST = os.environ.get('AWS_S3_HOST', 's3.amazonaws.com')
    AWS_S3_CUSTOM_DOMAIN = '%s.%s' % (AWS_STORAGE_BUCKET_NAME, AWS_S3_HOST)

    AWS_S3_OBJECT_PARAMETERS = {
        'CacheControl': 'max-age=86400',
    }

    STATIC_URL = "https://%s/%s/" % (AWS_S3_CUSTOM_DOMAIN, AWS_STATIC_LOCATION)

    MEDIA_URL = "https://%s/%s/" % (AWS_S3_CUSTOM_DOMAIN, AWS_PUBLIC_MEDIA_LOCATION)

    STATICFILES_STORAGE = 'openbook.storage_backends.S3StaticStorage'

    DEFAULT_FILE_STORAGE = 'openbook.storage_backends.S3PublicMediaStorage'

    PRIVATE_FILE_STORAGE = 'openbook.storage_backends.S3PrivateMediaStorage'

# ONE SIGNAL
ONE_SIGNAL_APP_ID = os.environ.get('ONE_SIGNAL_APP_ID')
ONE_SIGNAL_API_KEY = os.environ.get('ONE_SIGNAL_API_KEY')

# Peekalink

PEEKALINK_API_KEY = os.environ.get('PEEKALINK_API_KEY', None)

CORS_ALLOWED_ORIGINS = [
    "https://example.com",
    "http://161.35.161.138:80",
    "http://localhost:3000",
    "http://127.0.0.1:80"
]
CORS_ALLOW_ALL_ORIGINS = True
